contrib/vcl/source/vcl_interface.cc

```diff
@@ -0,0 +1,126 @@
+#include "contrib/vcl/source/vcl_interface.h"
+
+#include "source/common/network/address_impl.h"
+
+#include "contrib/vcl/source/vcl_io_handle.h"
+#include "vppcom.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Network {
+namespace Vcl {
+
+namespace {
+
+/**
+ * Max number of workers supported. Used in declaration of epoll_handles array
+ */
+const int MaxNumWorkers = 128;
+
+/**
+ * Max number of epoll events to drain from VCL per `vppcom_epoll_wait` call
+ */
+const int MaxNumEpollEvents = 128;
+
+/**
+ * Envoy worker epoll session handles by VCL worker index, i.e., `vppcom_worker_index()`. Each
+ * worker uses its respective handle to retrieve session events from VCL via `vppcom_epoll_wait()`.
+ */
+uint32_t epoll_handles[MaxNumWorkers];
+
+/**
+ * Mutex only used during VCL worker registration
+ */
+ABSL_CONST_INIT absl::Mutex wrk_lock(absl::kConstInit);
+
+using MqFileEventsMap = absl::flat_hash_map<int, Envoy::Event::FileEventPtr>;
+
+MqFileEventsMap& mqFileEventsMap() { MUTABLE_CONSTRUCT_ON_FIRST_USE(MqFileEventsMap); }
+
+void onMqSocketEvents(uint32_t flags) {
+  ASSERT((flags & (Event::FileReadyType::Read | Event::FileReadyType::Write)));
+  auto wrk_index = vppcom_worker_index();
+  VCL_LOG("events on worker {}", wrk_index);
+  struct epoll_event events[MaxNumEpollEvents];
+  int max_events = MaxNumEpollEvents, n_events;
+
+  while (max_events > 0) {
+    n_events = vppcom_epoll_wait(epoll_handles[wrk_index], events, max_events, 0);
```

Comment:
```suggestion
    int n_events = vppcom_epoll_wait(epoll_handles[wrk_index], events, max_events, 0);
```

And remove declaration above.